plugin/procfs/procfs.cc

```diff
@@ -0,0 +1,514 @@
+
+        if (realpath(fname.c_str(), real_file_path.data())) {
+          if (strncmp(SYS_NAME, real_file_path.data(), SYS_NAME_LEN) == 0 ||
+              strncmp(PROC_NAME, real_file_path.data(), PROC_NAME_LEN) == 0)
+            files_found.push_back(fname);
+        }
+        if (files_found.size() >= static_cast<size_t>(max_results)) break;
+      }
+    }
+  }
+
+  closedir(dir);
+  return;
+}
+
+static void limited_glob(const std::string &query, int max_results,
+                         int max_depth, std::vector<std::string> &files_found) {
+  if (max_results <= 0 || max_depth <= 0) return;
+
+  std::string path;
+  std::string pattern;
+  std::string rest;
+
+  std::string::size_type first_star = query.find_first_of("*?[");
+  std::string::size_type last_sep_before_star =
+      query.find_last_of("/", first_star);
+  std::string::size_type first_sep_after_star =
+      query.find_first_of("/", first_star);
+
+  if (last_sep_before_star == std::string::npos) {
+    path = std::string(".");
+    pattern = query;
+  } else if (first_sep_after_star == std::string::npos) {
+    path = std::string(query.begin(), query.begin() + last_sep_before_star);
+    pattern =
+        std::string(query.begin() + last_sep_before_star + 1, query.end());
+  } else {
+    path = std::string(query.begin(), query.begin() + last_sep_before_star);
+    pattern = std::string(query.begin() + last_sep_before_star + 1,
+                          query.begin() + first_sep_after_star);
+    rest = std::string(query.begin() + first_sep_after_star + 1, query.end());
+  }
+
+  if (rest.size() == 0) {
+    limited_glob_files(path, pattern, max_results, files_found);
+    return;
+  }
+
+  DIR *dir = opendir(path.c_str());
+  struct dirent *dir_entry;
+
+  if (!dir) return;
+
+  while ((dir_entry = readdir(dir)) != 0) {
+    if (!fnmatch(pattern.c_str(), dir_entry->d_name,
+                 FNM_CASEFOLD | FNM_NOESCAPE | FNM_PERIOD)) {
+      if (dir_entry->d_type == DT_DIR || dir_entry->d_type == DT_LNK ||
+          dir_entry->d_type == DT_UNKNOWN) {
+        std::string new_pattern = std::string(path)
+                                      .append("/")
+                                      .append(dir_entry->d_name)
+                                      .append("/")
+                                      .append(rest);
+        limited_glob(new_pattern, max_results - files_found.size(),
+                     max_depth - 1, files_found);
+      }
+    }
+  }
+
+  closedir(dir);
+  return;
+}
+
+static void fill_files_list(std::vector<std::string> &files) {
+  std::istringstream list(files_spec);
+
+  files.clear();
+  while (list) {
+    std::string path;
+    std::getline(list, path, ';');
+    if (path.rfind("/proc", 0) != 0 && path.rfind("/sys", 0) != 0) continue;
+
+    if (path.find_first_of("*?[") == std::string::npos) {
+      files.push_back(path);
+      continue;
+    }
+
+    limited_glob(path, MAX_FILES_PER_PATTERN, MAX_PATTERN_DEPTH, files);
```

Comment:
This is definitely out of scope of this PR, but there's an optimization (refactoring) possibility here, and we probably should at least create a jira ticket about it.

Right now, we traverse the proc/sys ft for every wildcard entry in the list. We probably want to do that only once, and compare it against all possible patterns.

Reply:
+1 see PS-7703